### ISSUE TITLE
Strange issue  with 3.1, background jobs, Heroku and HTML email attachments (#117)

### ISSUE THREAD

**Issue by jontyD** [mime-types/ruby-mime-types#127]
This seems to have re-occurred in rails 5.0.2, ruby 2.4.0
mime_types version 3.1

calling MIME::Types.of(filename) causes an error: undefined method `sort' for nil:NilClass
/app/vendor/bundle/ruby/2.4.0/gems/mime-types-3.1/lib/mime/types.rb:152:in `type_for'
/app/vendor/bundle/ruby/2.4.0/gems/mime-types-3.1/lib/mime/types/registry.rb:32:in `type_for'

filename is some_name.pdf and is definitely present at this point

This works locally including in a local delayed job but on heroku worker in a delayed job I get this error which seems like issue 117

NB dj running as a daemon

**Comment by halostatue** [comment]
… That‘s really weird. It’s pointed out something that I didn’t know (`[].inject(:+) => nil`), but the result array shouldn’t be empty because the extension index should be fully populated before this code gets run. This is going to take some time to figure out, and I’m afraid that I don’t see having that time in the next two weeks.

**Comment by johnsonj** [comment]
I'm hitting this trying to update [this code](https://github.com/johnsonj/cloud_controller_ng/blob/master/lib/cloud_controller/blobstore/fog/fog_client.rb#L63) from mime-types 2.6.2 to 3.1. Will update if I find more info.

**Comment by FX-HAO** [comment]
+ 1

**Comment by johnsonj** [comment]
In our case the issue isn't that @extensions_index is nil, it's that we were passing in a bad parameter that it was previously quiet about.

The calling code essentially did this: 

```ruby
  MIME::Types.of(Tempfile.new).first.try(:content_type)
```

Which is used to create an empty array inside of [type_for](https://github.com/mime-types/ruby-mime-types/blob/master/lib/mime/types.rb#L149)
```ruby
  a = Array(Tempfile.new) # => []
  a.flat_map { |fn| puts 'never called!'; fn } # => []
  a.compact # => []
  a.inject(:+) # => nil
```

So ruby learnings from this bug:
 - `[].inject(:+) => nil`
 - `[Tempfile.new]` ***!=*** `Array.new(Tempfile.new)`

**Comment by halostatue** [comment]
This is because `Array(Tempfile.new)` calls `Tempfile#to_a`, delegated from `File#to_a` via `Enumerable#to_a`.

Sounds like I might need to go back to `[arg].flatten` instead of `Array(arg)`.

**Comment by halostatue** [comment]
Resolved with edf3cd7.
